### PR TITLE
Refactor global verbosity value to parameters

### PR DESCRIPTION
**Pull request by krzyzanowskim**
Refactor global `TSCUtils.verbosity` value to functions parameter

### Motivation:

While it's okayish to maintain global `TSCUtils.verbosity` for CLI tool, it's not really for library API. I believe it's better to provide a configurable verbosity parameter for library public API.

### Modifications:

replace `TSCUtils.verbosity` with function parameter that default to global `TSCUtils.verbosity`

### Result:

API verbosity setup is more granulated.

### CODE REVIEW

File: Sources/Build/BuildOperation.swift
Comment (tomerd) [inline Sources/Build/BuildOperation.swift:52]: wdyt about storing `Verbosity` instead of `Bool` as the member, and adding a "isVerbose" computed property that is calculated off of that enum value?

File: Sources/PackageLoading/ManifestLoader.swift
Comment (tomerd) [inline Sources/PackageLoading/ManifestLoader.swift:128]: I'd rather not have a default set here to make sure the call-site make a conscious decision. For the unit test we can have an short-form via extension

File: Sources/PackageLoading/ManifestLoader.swift
Comment (tomerd) [inline Sources/PackageLoading/ManifestLoader.swift:270]: this API is deprecated so we should not add argument here, just pass in  TSCUtility.verbosity as default in L282

File: Sources/PackageLoading/ManifestLoader.swift
Comment (tomerd) [inline Sources/PackageLoading/ManifestLoader.swift:328]: not: align this to the back

File: Sources/XCBuildSupport/XcodeBuildSystem.swift
Comment (tomerd) [inline Sources/XCBuildSupport/XcodeBuildSystem.swift:55]: as in the other spot, I prefer passing the full Verbosity value and not having a default so call-sites make conscious decision

**Review by tomerd — CHANGES_REQUESTED** [review]
thanks for putting together @krzyzanowskim, looks good, some ideas inline

File: Sources/Build/BuildOperation.swift
Comment (krzyzanowskim) [reply Sources/Build/BuildOperation.swift:52]: it's akin to what's in `ToolWorkspaceDelegate`, would you prefer to change there as well?

**Comment by krzyzanowskim** [comment]
@tomerd back to you. After I removed the default value for verbosity, it surfaced even more places where `verbosity` parameter is added.

File: Sources/Build/BuildOperation.swift
Comment (tomerd) [inline Sources/Build/BuildOperation.swift:208]: should buildDelegate also take verbosity instead of boolean?

File: Sources/Commands/APIDigester.swift
Comment (tomerd) [inline Sources/Commands/APIDigester.swift:116]: should we define a verbosity variable and share it between test two calls?

File: Sources/Commands/APIDigester.swift
Comment (tomerd) [inline Sources/Commands/APIDigester.swift:183]: not sure I am reading this right... is it comparing the global value that is coming from TSCUtility? should this be self.verbosity?

File: Sources/Commands/SwiftTool.swift
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:384]: feels like we should create a member variable here instead of setting TSCUtility.verbosity, no?

File: Sources/Commands/SwiftTool.swift
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:534]: should this use options.verbosity?

File: Sources/Commands/SwiftTool.swift
Comment (tomerd) [reply Sources/Commands/SwiftTool.swift:384]: or it this designed so that some downstream code that relies on the global continues to work?

File: Sources/Commands/SwiftTool.swift
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:642]: options.verbosity?

File: Sources/Commands/SwiftTool.swift
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:661]: options.verbosity?

File: Sources/Commands/SwiftTool.swift
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:670]: options.verbosity?

File: Sources/Commands/SymbolGraphExtract.swift
Comment (tomerd) [inline Sources/Commands/SymbolGraphExtract.swift:62]: should this take verbosity instead of boolean?

File: Sources/SPMBuildCore/BuildParameters.swift
Comment (tomerd) [inline Sources/SPMBuildCore/BuildParameters.swift:196]: do we need the default?

File: Sources/Build/BuildOperation.swift
Comment (krzyzanowskim) [reply Sources/Build/BuildOperation.swift:208]: you think so? maybe. I bet there are more places where `isVerbose` is in use

File: Sources/Commands/APIDigester.swift
Comment (krzyzanowskim) [reply Sources/Commands/APIDigester.swift:116]: This is how it behaves until now anyway, doesn't it? I mean, the global verbosity is what is used.

File: Sources/Commands/SwiftTool.swift
Comment (krzyzanowskim) [reply Sources/Commands/SwiftTool.swift:384]: yes, I didn't replace who it works currently so I put the global value that is managed somewhere in the tool flow.

File: Sources/Commands/SwiftTool.swift
Comment (krzyzanowskim) [reply Sources/Commands/SwiftTool.swift:534]: `ToolWorkspaceDelegate` got the `isVerbose` as an initializer parameter so I guess this is more important than options. In general, I agree that this should be in options however, that's not where the SwiftTool is yet, and I'd prefer to split that in another PR. As of now, the global variable is still in use by the CLI code.

File: Sources/Commands/APIDigester.swift
Comment (krzyzanowskim) [reply Sources/Commands/APIDigester.swift:183]: there's no `self.verbosity`, it uses `TSCUtiliry.verbosity` all the way as it used to. What we can see now is how not perfect this global is.

File: Sources/Commands/SymbolGraphExtract.swift
Comment (krzyzanowskim) [reply Sources/Commands/SymbolGraphExtract.swift:62]: `Process` is from `swift-tools-support-core` package though. Can't change it now.

File: Sources/SPMBuildCore/BuildParameters.swift
Comment (krzyzanowskim) [reply Sources/SPMBuildCore/BuildParameters.swift]: no, removed

File: Sources/Commands/SwiftTool.swift
Comment (krzyzanowskim) [reply Sources/Commands/SwiftTool.swift:642]: see https://github.com/apple/swift-package-manager/pull/3216#discussion_r562598849

File: Sources/Commands/SwiftTool.swift
Comment (krzyzanowskim) [reply Sources/Commands/SwiftTool.swift:661]: see https://github.com/apple/swift-package-manager/pull/3216#discussion_r562598849

File: Sources/Commands/SwiftTool.swift
Comment (krzyzanowskim) [reply Sources/Commands/SwiftTool.swift:670]: see https://github.com/apple/swift-package-manager/pull/3216#discussion_r562598849

**Comment by krzyzanowskim** [comment]
@tomerd thanks for the feedback. My goal in this PR is to make only a library part free of global `TSCUtility.verbosity`. I agree it's unfortunate that the code relies on the global variable from the `TSCUtility` module. Ideally, the `TSCUtility.verbosity` should be removed. I'd rather do it in another task.

**Comment by tomerd** [comment]
> @tomerd thanks for the feedback. My goal in this PR is to make only a library part free of global TSCUtility.verbosity. I agree it's unfortunate that the code relies on the global variable from the TSCUtility module. Ideally, the TSCUtility.verbosity should be removed. I'd rather do it in another task.

thank you @krzyzanowskim  for taking this on, i agree removing TSCUtility.verbosity completely could be a separate task, but we need to make sure we dont end up in too much of a mixed state otherwise it would be very hard to reason about the code

File: Sources/Build/BuildOperation.swift
Comment (tomerd) [reply Sources/Build/BuildOperation.swift:208]: this one is less critical

File: Sources/Commands/APIDigester.swift
Comment (tomerd) [reply Sources/Commands/APIDigester.swift:183]: right, we should do away with the global imo. but this one used to be "verbosity", where did it come from before?

File: Sources/Commands/SymbolGraphExtract.swift
Comment (tomerd) [reply Sources/Commands/SymbolGraphExtract.swift:62]: okay

**Comment by tomerd** [comment]
cc @abertelrud @neonichu for opinions on approach

File: Sources/Commands/APIDigester.swift
Comment (krzyzanowskim) [reply Sources/Commands/APIDigester.swift:183]: It was still from the `TSCUtility`, just used without module prefix. I made it explicit.

Comment (tomerd) [reply Sources/Commands/APIDigester.swift:183]: okay that makes total sense

**Comment by neonichu** [comment]
If we see this as a stepping stone to removing `TSCUtility.verbosity`, I think the approach is fine.

**Comment by abertelrud** [comment]
I agree with @neonichu that incremental cleanup is fine as long as the goal is to remove the global (possibly in another PR).

**Comment by krzyzanowskim** [comment]
It's been a while without approval and many changes happened in the meantime. I don't think I can invest more spare time to keep it updated. Feel free to take it over, or discard and leave the global verbosity in the code if that's what you think is best right now.

**Comment by tomerd** [comment]
@krzyzanowskim apologies for the delay on this. we have been waiting for the [5.5 branching](https://forums.swift.org/t/swift-5-5-release-process/45644) to take changes that should land post 5.5. we tried to mark those with the "next" label but failed to do so on this one.

given @abertelrud and @neonichu opinion here, I think we can approve this and merge after the branching. obviously this need rebasing which I think is what you were pointing out

**Comment by krzyzanowskim** [comment]
> obviously this need rebasing which I think is what you were pointing out

yea, it a bit unfortunate for me personally (subjective stand) to not know for how long need to maintain PR to get it in. It's fine though. If anyone has spare cycles, feel free to take it over, and rebase, is all I wanted to say.

**Comment by krzyzanowskim** [comment]
> obviously this need rebasing which I think is what you were pointing out

right, it's just a bit unfortunate for me personally to not know for how long need to maintain PR before it gets in. It's fine though. If anyone has spare cycles, feel free to take it over, and rebase, which is all I wanted to say.

**Comment by tomerd** [comment]
thanks for your patience @krzyzanowskim. 5.5 was just branched this weekend so we are finally ready to take this. would you like to rebase, or do you prefer someone take this over / close and re-do, etc?

**Comment by tomerd** [comment]
@krzyzanowskim  closing for now as this is so far off mainline. will track on our refactoring list tho, as this is a good cleanup.

**Pull request closed without merging**
